synapse/storage/database.py

```diff
@@ -672,7 +672,15 @@ def new_transaction(
             f = cast(types.FunctionType, func)  # type: ignore[redundant-cast]
             if f.__closure__:
                 for i, cell in enumerate(f.__closure__):
-                    if inspect.isgenerator(cell.cell_contents):
+                    try:
+                        contents = cell.cell_contents
+                    except ValueError:
+                        # cell.cell_contents can raise if the "cell" is empty,
+                        # which indicates that the variable is currently
+                        # unbound.
```

Comment:
Argh, sorry missed this comment.

It's because we had the following construct in the new code:

```python
if a:
    b = set()

def foo():
    if a:
        b.add(1)

foo()
```

Where if `a` is false, `b` is unbound and so `cell.cell_contents` raises a `ValueError: cell is empty`